model-engine/model_engine_server/inference/batch_inference/vllm_batch.py

```diff
@@ -0,0 +1,195 @@
+    if content is None:
+        with open(request.input_data_path, "r") as f:
+            content = CreateBatchCompletionsRequestContent.parse_raw(f.read())
+
+    model = (
+        "./model_weights" if request.model_config.checkpoint_path else request.model_config.model
+    )
+
+    engine_args = AsyncEngineArgs(
+        model=model,
+        quantization=request.model_config.quantize,
+        tensor_parallel_size=request.model_config.num_shards,
+        seed=request.model_config.seed or 0,
+    )
+
+    llm = AsyncLLMEngine.from_engine_args(engine_args)
+
+    # Add the requests to the engine.
+    sampling_params = SamplingParams(
+        max_tokens=content.max_new_tokens,
+        temperature=content.temperature,
+        stop=content.stop_sequences,
+        logprobs=1 if content.return_token_log_probs else None,
+        presence_penalty=content.presence_penalty or 0.0,
+        frequency_penalty=content.frequency_penalty or 0.0,
+        top_k=content.top_k or -1,
+        top_p=content.top_p or 1.0,
+    )
+
+    results_generators = []
+    prompts_per_pod = len(content.prompts) // request.data_parallelism
+    for prompt in content.prompts[prompts_per_pod * job_index : prompts_per_pod * (job_index + 1)]:
+        request_id = random_uuid()
+        results_generator = await llm.add_request(
+            request_id, prompt, sampling_params, None, time.monotonic()
+        )
+        results_generators.append(results_generator)
+
+    outputs = []
+    for generator in results_generators:
+        last_output_text = ""
+        tokens = []
+        async for request_output in generator:
+            token_text = request_output.outputs[-1].text[len(last_output_text) :]
+            log_probs = request_output.outputs[0].logprobs[-1] if sampling_params.logprobs else None
+            last_output_text = request_output.outputs[-1].text
+
+            if content.return_token_log_probs:
+                tokens.append(TokenOutput(token=token_text, log_prob=list(log_probs.values())[0]))
+
+        num_prompt_tokens = len(request_output.prompt_token_ids)
+        num_completion_tokens = len(request_output.outputs[0].token_ids)
+
+        output = CompletionOutput(
+            text=request_output.outputs[0].text,
+            num_prompt_tokens=num_prompt_tokens,
+            num_completion_tokens=num_completion_tokens,
+        )
+        if content.return_token_log_probs:
+            output.tokens = tokens
+
+        outputs.append(output.dict())
+
+    if request.data_parallelism == 1:
+        with open(request.output_data_path, "w") as f:
+            f.write(json.dumps(outputs))
+    else:
+        chunk_file = f"{request.output_data_path}.{job_index}"
+        with open(chunk_file, "w") as f:
+            f.write(json.dumps(outputs))
+        if job_index == 0:
+            wait_for_all_chunks(request)
+            combine_all_chunks(request)
+            if request.output_data_path.startswith("s3://"):
```

Comment:
Should we have this in a `finally`?

Reply:
i think this "finally" should be at end of job. otherwise if worker 0 failed early, it may not be able to remove the other output chunks.
for now i think it's fine to not remove and leave some traces for debugging.